### ISSUE TITLE
S3 Rename - Preserve ACL settings

### ISSUE THREAD

**Issue by caoimhinb** [servocoder/RichFilemanager-PHP#6]
When you rename a file, it loses its ACL settings.  If you have defaultAcl set to ACL_PUBLIC_READ, the file first gets uploaded with that setting but then when you rename it, that gets set to private [so the preview urls are then broken].  I haven't found a workaround for this yet, but I think it may involve changing StorageHelper.php so that instead of standard php 'rename' it should do an S3 copy + delete?

**Comment by servocoder** [comment]
Thanks for the report, I will investigate this case.

**Comment by caoimhinb** [comment]
I've made some modifications to get this to work at least for what I needed.  
Rename will copy and delete using the defaultACL (so if there are special existing settings on some files, they will still be replaced with defaultACL).

Note : I have only tested with S3 for file+image storage and local storage for the thumbnails.  

I changed 3 files:
AwsS3API.php, from line 234 - use dynamicPath, and call renameRecursive for S3 objects, local thumbs keep using php 'rename':
```php
$S3Source=$modelOld->getDynamicPath();
        $S3Destination=$modelNew->getDynamicPath();

        if ($this->storage->renameRecursive($S3Source, $S3Destination)) {
            Log::info('renamed "' . $S3Source . '" to "' . $S3Destination . '"');

            if ($modelThumbOld->isExists) {
                if ($this->storage->config('images.thumbnail.useLocalStorage')) {
                    rename($modelThumbOld->pathAbsolute, $modelThumbNew->pathAbsolute);
                } else {
                    $this->storage->renameRecursive($modelThumbOld->getDynamicPath(), $modelThumbNew->getDynamicPath());
                }
            }
        }
```

S3/Storage.php, in renameRecursive, stripped out and just uses the existing rename method from StorageHelper.php:
```php
 public function renameRecursive($source, $target)
    {

        return $this->s3->rename($source, $target); 
   }
```

S3/StorageHelper.php:
```php
public function rename($key, $destination)
    {
        $key=ltrim($key, '/');
        $destination=ltrim($destination, '/');
        $bucket = $this->bucket;
        $isDir = is_dir("s3://{$bucket}/{$key}");

        if($isDir) {
            $result = $this->getList($key);

            if (!isset($result['Contents'])) {
                return false;
            }

            foreach ($result['Contents'] as $object) {
                $newPath = str_replace($key, $destination, $object['Key']);
                if ($this->copy($object['Key'], $newPath)) {
                    $this->delete($object['Key']);
                }

            }
        }
        else{
            if($this->copy($key, $destination)){
                return $this->delete($key);
            }
            else{
                return false;
            }
        }

        return true;
    }
```

**Comment by caoimhinb** [comment]
I've found the exact same problem exists with copy and move actions on S3.  
The expected behavior I think would be for copied/moved/renamed files to retain their existing ACL settings - which is what would happen with Local File manipulation.  For that to happen, an extra S3 call would be needed to get current settings for the object and use that for the destination too.

**Comment by servocoder** [comment]
You are right, the problem exists for copy/move/rename actions.
I have made a review of code and found few ways to solve the problem.

First we can force file/folder ACL rules for all actions, no matter what ACL it originally had. For example you could have a file with "public" ACL, but if there is "private" ACL set in the settings it will be turned to "private" after copy/move/rename action. It's a rough way.

We can go further and check ACL rules for source(original) file  before performing copy/move/rename actions and apply it to the desctination file. It will require extra GET request to S3 to fetch ACL rules for each file/folder. Also ACL rules more complicated then just private/public and I'm not sure it will be possible to keep the complete ACL rules for the destination file. But this seems quite possible to keep "private"/"public" ACL of original file.

I believe we can add another option to s3 config file so user can determine the behavior of ACL rules: FORCE vs INHERIT. What do you think?

**Comment by caoimhinb** [comment]
Yes I completely agree that the best possible way would be to allow the user to specify which way they prefer.  Because it involves extra GET calls, preserving the ACL will be slower (especially renaming a folder containing hundreds of files for example).  So, FORCE may be better for people looking for efficiency, but INHERIT is better for security and compatibility reasons.

**Comment by servocoder** [comment]
Glad to hear we have the consent.
Ok, I will be busy this week and back to this issue in a week or more.

**Comment by servocoder** [comment]
Implemented in v1.1.0, any feedback is greatly appreciated
